File: solutions/csharp/pig-latin/1/PigLatin.cs
using System;
using System.Collections.Generic;
using System.Linq;

// ReSharper disable once CheckNamespace
public static class PigLatin
{
	private const StringComparison _invariantCulture = StringComparison.InvariantCulture;
	private const char _separator = ' ';

	private static readonly HashSet<char> _vowels =
		new()
		{
			'a',
			'e',
			'i',
			'o',
			'u',
		};

	public static string Translate(string text)
	{
		if (string.IsNullOrWhiteSpace(text))
		{
			throw new ArgumentException($"'{nameof(text)}' cannot be null or whitespace.", nameof(text));
		}

		IEnumerable<string?> translatedWords = text.Split(_separator)
			.Select(TranslateWord);
		return string.Join(" ", translatedWords);
	}

	private static string TranslateWord(string text)
		=> text.StartsWithVowel()
			? text.AddAy()
			: text.TranslateWhenStartingWithConsonant();

	private static string TranslateWhenStartingWithConsonant(this string word)
	{
		int firstVowel = word.FindFirstVowel();
		return word.StartsWithQu()
			? $"{word[2..]}qu".AddAy()
			: word.HasQuPrecededByConsonant()
				? $"{word[3..]}{word[0]}qu".AddAy()
				: word.IsTwoLetterYWord()
					? $"{word[1]}{word[0]}".AddAy()
					: word.HasYConsonant(firstVowel)
						? word.TranslateWhenContainingYVowel()
						: $"{word[firstVowel..]}{word[..firstVowel]}".AddAy();
	}

	private static string TranslateWhenContainingYVowel(this string word)
	{
		int indexOfY = word.IndexOf("y", _invariantCulture);
		return $"{word[indexOfY..]}{word[..indexOfY]}".AddAy();
	}

	private static bool StartsWithVowel(this string word)
		=> _vowels.Contains(word[0])
		   || ((word.Length > 2)
			   && (
				   word.StartsWith("xr", _invariantCulture)
				   || word.StartsWith("yt", _invariantCulture)
			   )
		   );

	private static int FindFirstVowel(this string word)
		=> word.Select((c, index) => (c, index))
			.FirstOrDefault(pair => _vowels.Contains(pair.c))
			.index;

	private static bool IsTwoLetterYWord(this string word)
		=> (word.Length == 2) && word.Contains('y', _invariantCulture);

	private static bool HasYConsonant(this string word, int firstVowel)
	{
		int indexOfY = word.IndexOf("y", _invariantCulture);
		return IsConsonant(word[0])
			   && (indexOfY > 1)
			   && ((firstVowel == 0)
				   || (firstVowel > indexOfY));
	}

	private static bool IsConsonant(this char value)
		=> !_vowels.Contains(value);

	private static bool StartsWithQu(this string word)
		=> word.StartsWith("qu", _invariantCulture);

	private static bool HasQuPrecededByConsonant(this string word)
		=> word[0]
			   .IsConsonant()
		   && (word[1] == 'q')
		   && (word[2] == 'u');

	private static string AddAy(this string value)
		=> $"{value}ay";
}

File: solutions/python/pig-latin/2/pig_latin.py
VOWELS = "aeiou"
AY_ENDING = "ay"

def is_vowel(character, index):
    return character in VOWELS or (character == "y" and index > 0)

def get_translation_start(word):
    if _starts_with_vowel(word) or _starts_with_special_vowel_sound(word):
        return 0

    prefix_length = _get_consonant_prefix_length(word)
    if word[prefix_length - 1:prefix_length + 1] == "qu":
        return prefix_length  + 1

    return prefix_length

def _starts_with_vowel(word):
    return word[0] in VOWELS

def _starts_with_special_vowel_sound(word):
    return word.startswith(("xr", "yt"))

def _get_consonant_prefix_length(word):
    index = 0
    while (
            index < len(word)
           and not is_vowel(word[index], index)
    ):
        index += 1

    return index

def translate_word(word):
    translation_start = get_translation_start(word)

    if translation_start == 0:
        return word + AY_ENDING

    beginning = word[:translation_start]
    remainder = word[translation_start:]
    return remainder + beginning + AY_ENDING

def translate(text):
    translated_words = []
    for word in text.split():
        translated_words.append(translate_word(word))

    return " ".join(translated_words)